## Microsoft book online describes the APPLY keyword allow a query to invoke a table value function.  This is only one of the uses of the APPLY keyword. 

## First lets look at the basic syntax using Cross Apply

#

<h1> SELECT </br>
 * </br>
FROM Table AS T</br>
CROSS APPLY </br>
[dbo].[exampleTVF] (T.fk) AS CA</br>

<h1> SELECT </br>
 * </br>
FROM Table AS T</br>
OUTER APPLY </br>
[dbo].[exampleTVF] (T.fk) AS CA</br>

In [2]:
SELECT 
*
FROM Table AS T
CROSS APPLY 
[dbo].[exampleTVF](T.fk) AS CA


: Msg 156, Level 15, State 1, Line 3
Incorrect syntax near the keyword 'Table'.

Total execution time: 00:00:00.015

## In what way are CROSS and OUTER different from each other? 

## CROSS APPLY - can be considered the equivalent of an inner join it will only return a record if the recordset has records from both tables 

## OUTER APPLY - can be considered the equivalent of an left join it will return all the records for the left table even if the function or query does not return any records.

Now let's look at what this means in practice

![sad panda](./Figure1.jpg "Figure 1 - returns a list of products from the Product table")</br>
Figure 1 - returns a list of products from the Product table


In [3]:
SELECT [P].[ProductID],
       [P].[Name],
       [P].[ProductNumber]
FROM [Production].[Product] AS [P]
WHERE [P].[FinishedGoodsFlag] = 1;

(295 rows affected)

Total execution time: 00:00:00.457

ProductID Name ProductNumber 680 HL Road Frame - Black, 58 FR-R92B-58 706 HL Road Frame - Red, 58 FR-R92R-58 707 Sport-100 Helmet, Red HL-U509-R 708 Sport-100 Helmet, Black HL-U509 709 Mountain Bike Socks, M SO-B909-M 710 Mountain Bike Socks, L SO-B909-L 711 Sport-100 Helmet, Blue HL-U509-B 712 AWC Logo Cap CA-1098 713 Long-Sleeve Logo Jersey, S LJ-0192-S 714 Long-Sleeve Logo Jersey, M LJ-0192-M 715 Long-Sleeve Logo Jersey, L LJ-0192-L 716 Long-Sleeve Logo Jersey, XL LJ-0192-X 717 HL Road Frame - Red, 62 FR-R92R-62 718 HL Road Frame - Red, 44 FR-R92R-44 719 HL Road Frame - Red, 48 FR-R92R-48 720 HL Road Frame - Red, 52 FR-R92R-52 721 HL Road Frame - Red, 56 FR-R92R-56 722 LL Road Frame - Black, 58 FR-R38B-58 723 LL Road Frame - Black, 60 FR-R38B-60 724 LL Road Frame - Black, 62 FR-R38B-62 725 LL Road Frame - Red, 44 FR-R38R-44 726 LL Road Frame - Red, 48 FR-R38R-48 727 LL Road Frame - Red, 52 FR-R38R-52 728 LL Road Frame - Red, 58 FR-R38R-58 729 LL Road Frame - Red, 60 FR-R38R-60 730 LL Road Frame - Red, 62 FR-R38R-62 731 ML Road Frame - Red, 44 FR-R72R-44 732 ML Road Frame - Red, 48 FR-R72R-48 733 ML Road Frame - Red, 52 FR-R72R-52 734 ML Road Frame - Red, 58 FR-R72R-58 735 ML Road Frame - Red, 60 FR-R72R-60 736 LL Road Frame - Black, 44 FR-R38B-44 737 LL Road Frame - Black, 48 FR-R38B-48 738 LL Road Frame - Black, 52 FR-R38B-52 739 HL Mountain Frame - Silver, 42 FR-M94S-42 740 HL Mountain Frame - Silver, 44 FR-M94S-44 741 HL Mountain Frame - Silver, 48 FR-M94S-52 742 HL Mountain Frame - Silver, 46 FR-M94S-46 743 HL Mountain Frame - Black, 42 FR-M94B-42 744 HL Mountain Frame - Black, 44 FR-M94B-44 745 HL Mountain Frame - Black, 48 FR-M94B-48 746 HL Mountain Frame - Black, 46 FR-M94B-46 747 HL Mountain Frame - Black, 38 FR-M94B-38 748 HL Mountain Frame - Silver, 38 FR-M94S-38 749 Road-150 Red, 62 BK-R93R-62 750 Road-150 Red, 44 BK-R93R-44 751 Road-150 Red, 48 BK-R93R-48 752 Road-150 Red, 52 BK-R93R-52 753 Road-150 Red, 56 BK-R93R-56 754 Road-450 Red, 58 BK-R68R-58 755 Road-450 Red, 60 BK-R68R-60 756 Road-450 Red, 44 BK-R68R-44 757 Road-450 Red, 48 BK-R68R-48 758 Road-450 Red, 52 BK-R68R-52 759 Road-650 Red, 58 BK-R50R-58 760 Road-650 Red, 60 BK-R50R-60 761 Road-650 Red, 62 BK-R50R-62 762 Road-650 Red, 44 BK-R50R-44 763 Road-650 Red, 48 BK-R50R-48 764 Road-650 Red, 52 BK-R50R-52 765 Road-650 Black, 58 BK-R50B-58 766 Road-650 Black, 60 BK-R50B-60 767 Road-650 Black, 62 BK-R50B-62 768 Road-650 Black, 44 BK-R50B-44 769 Road-650 Black, 48 BK-R50B-48 770 Road-650 Black, 52 BK-R50B-52 771 Mountain-100 Silver, 38 BK-M82S-38 772 Mountain-100 Silver, 42 BK-M82S-42 773 Mountain-100 Silver, 44 BK-M82S-44 774 Mountain-100 Silver, 48 BK-M82S-48 775 Mountain-100 Black, 38 BK-M82B-38 776 Mountain-100 Black, 42 BK-M82B-42 777 Mountain-100 Black, 44 BK-M82B-44 778 Mountain-100 Black, 48 BK-M82B-48 779 Mountain-200 Silver, 38 BK-M68S-38 780 Mountain-200 Silver, 42 BK-M68S-42 781 Mountain-200 Silver, 46 BK-M68S-46 782 Mountain-200 Black, 38 BK-M68B-38 783 Mountain-200 Black, 42 BK-M68B-42 784 Mountain-200 Black, 46 BK-M68B-46 785 Mountain-300 Black, 38 BK-M47B-38 786 Mountain-300 Black, 40 BK-M47B-40 787 Mountain-300 Black, 44 BK-M47B-44 788 Mountain-300 Black, 48 BK-M47B-48 789 Road-250 Red, 44 BK-R89R-44 790 Road-250 Red, 48 BK-R89R-48 791 Road-250 Red, 52 BK-R89R-52 792 Road-250 Red, 58 BK-R89R-58 793 Road-250 Black, 44 BK-R89B-44 794 Road-250 Black, 48 BK-R89B-48 795 Road-250 Black, 52 BK-R89B-52 796 Road-250 Black, 58 BK-R89B-58 797 Road-550-W Yellow, 38 BK-R64Y-38 798 Road-550-W Yellow, 40 BK-R64Y-40 799 Road-550-W Yellow, 42 BK-R64Y-42 800 Road-550-W Yellow, 44 BK-R64Y-44 801 Road-550-W Yellow, 48 BK-R64Y-48 802 LL Fork FK-1639 803 ML Fork FK-5136 804 HL Fork FK-9939 805 LL Headset HS-0296 806 ML Headset HS-2451 807 HL Headset HS-3479 808 LL Mountain Handlebars HB-M243 809 ML Mountain Handlebars HB-M763 810 HL Mountain Handlebars HB-M918 811 LL Road Handlebars HB-R504 812 ML Road Handlebars HB-R720 813 HL Road Handlebars HB-R956 814 ML Mou

## The query in Figure 1 returns 295 rows, and is a list of products from the [Production].[Product] table.  Next step is to create a function called [fn_TotalSoldToDate], this will take a ProductID and return the total number items ordered from the [Sales].[SalesOrderDetail] table see Figure 2.  For the  Nothing complicated so far.

## Now we are going to create a function called 'fn_TotalSoldToDate'

![sad panda](./Figure2.jpg "Figure 2 - script to create table value function [fn_TotalSoldToDate]")</br>
Figure 2 - script to create table value function [fn_TotalSoldToDate]

## Note 1
## _Eagled eyed readers will have noted that the example used for this table valued function only returns one row.  Which is **not** how table value functions are intended to be used.  The reason that I have used a table valued function is for demonstration purposes only.  So yes it’s not best practice._


In [4]:
IF OBJECT_ID (N'fn_TotalSoldToDate', N'IF') IS NOT NULL
  DROP FUNCTION dbo.fn_TotalSoldToDate
GO
CREATE FUNCTION fn_TotalSoldToDate (@ProductID int)
RETURNS TABLE
AS
RETURN
(
SELECT [SODT].ProductID,
       SUM([SODT].OrderQty) AS TotalOrdered
FROM [Sales].[SalesOrderDetail] AS [SODT]
GROUP BY ProductID
HAVING
    [SODT].ProductID = @ProductID
)
GO

Commands completed successfully.

Commands completed successfully.

Total execution time: 00:00:00.029

## Now the function has been created let's use the table value function with the query shown in figure 1 to return the total number of items sold for the products listed.

![sad panda](./Figure3.jpg "Figure 3 - Using a CROSS APPLY with the [fn_TotalSoldToDate] to return the total number sold for each productID")</br>
Figure 1 - returns a list of products from the Product table

In [5]:
SELECT [P].[ProductID],
       [P].[Name],
       [P].[ProductNumber],
	   [PS].[TotalOrdered]
FROM [Production].[Product] AS [P]
CROSS APPLY [dbo].[fn_TotalSoldToDate]([P].ProductID) AS [PS]
WHERE [P].[FinishedGoodsFlag] = 1;

(266 rows affected)

Total execution time: 00:00:00.472

ProductID Name ProductNumber TotalOrdered 707 Sport-100 Helmet, Red HL-U509-R 6266 708 Sport-100 Helmet, Black HL-U509 6532 709 Mountain Bike Socks, M SO-B909-M 1107 710 Mountain Bike Socks, L SO-B909-L 90 711 Sport-100 Helmet, Blue HL-U509-B 6743 712 AWC Logo Cap CA-1098 8311 713 Long-Sleeve Logo Jersey, S LJ-0192-S 429 714 Long-Sleeve Logo Jersey, M LJ-0192-M 3636 715 Long-Sleeve Logo Jersey, L LJ-0192-L 6592 716 Long-Sleeve Logo Jersey, XL LJ-0192-X 2980 717 HL Road Frame - Red, 62 FR-R92R-62 485 718 HL Road Frame - Red, 44 FR-R92R-44 486 719 HL Road Frame - Red, 48 FR-R92R-48 109 722 LL Road Frame - Black, 58 FR-R38B-58 940 723 LL Road Frame - Black, 60 FR-R38B-60 129 725 LL Road Frame - Red, 44 FR-R38R-44 991 726 LL Road Frame - Red, 48 FR-R38R-48 671 727 LL Road Frame - Red, 52 FR-R38R-52 100 729 LL Road Frame - Red, 60 FR-R38R-60 996 730 LL Road Frame - Red, 62 FR-R38R-62 697 732 ML Road Frame - Red, 48 FR-R72R-48 250 733 ML Road Frame - Red, 52 FR-R72R-52 90 736 LL Road Frame - Black, 44 FR-R38B-44 238 738 LL Road Frame - Black, 52 FR-R38B-52 1581 739 HL Mountain Frame - Silver, 42 FR-M94S-42 347 741 HL Mountain Frame - Silver, 48 FR-M94S-52 173 742 HL Mountain Frame - Silver, 46 FR-M94S-46 649 743 HL Mountain Frame - Black, 42 FR-M94B-42 1181 744 HL Mountain Frame - Black, 44 FR-M94B-44 17 745 HL Mountain Frame - Black, 48 FR-M94B-48 131 747 HL Mountain Frame - Black, 38 FR-M94B-38 659 748 HL Mountain Frame - Silver, 38 FR-M94S-38 1206 749 Road-150 Red, 62 BK-R93R-62 600 750 Road-150 Red, 44 BK-R93R-44 437 751 Road-150 Red, 48 BK-R93R-48 493 752 Road-150 Red, 52 BK-R93R-52 458 753 Road-150 Red, 56 BK-R93R-56 664 754 Road-450 Red, 58 BK-R68R-58 582 755 Road-450 Red, 60 BK-R68R-60 350 756 Road-450 Red, 44 BK-R68R-44 346 757 Road-450 Red, 48 BK-R68R-48 156 758 Road-450 Red, 52 BK-R68R-52 710 759 Road-650 Red, 58 BK-R50R-58 618 760 Road-650 Red, 60 BK-R50R-60 2232 761 Road-650 Red, 62 BK-R50R-62 1896 762 Road-650 Red, 44 BK-R50R-44 2254 763 Road-650 Red, 48 BK-R50R-48 1888 764 Road-650 Red, 52 BK-R50R-52 1112 765 Road-650 Black, 58 BK-R50B-58 1871 766 Road-650 Black, 60 BK-R50B-60 1100 767 Road-650 Black, 62 BK-R50B-62 644 768 Road-650 Black, 44 BK-R50B-44 1080 769 Road-650 Black, 48 BK-R50B-48 644 770 Road-650 Black, 52 BK-R50B-52 2270 771 Mountain-100 Silver, 38 BK-M82S-38 642 772 Mountain-100 Silver, 42 BK-M82S-42 593 773 Mountain-100 Silver, 44 BK-M82S-44 601 774 Mountain-100 Silver, 48 BK-M82S-48 505 775 Mountain-100 Black, 38 BK-M82B-38 682 776 Mountain-100 Black, 42 BK-M82B-42 634 777 Mountain-100 Black, 44 BK-M82B-44 678 778 Mountain-100 Black, 48 BK-M82B-48 616 779 Mountain-200 Silver, 38 BK-M68S-38 2394 780 Mountain-200 Silver, 42 BK-M68S-42 2234 781 Mountain-200 Silver, 46 BK-M68S-46 2216 782 Mountain-200 Black, 38 BK-M68B-38 2977 783 Mountain-200 Black, 42 BK-M68B-42 2664 784 Mountain-200 Black, 46 BK-M68B-46 2111 785 Mountain-300 Black, 38 BK-M47B-38 684 786 Mountain-300 Black, 40 BK-M47B-40 776 787 Mountain-300 Black, 44 BK-M47B-44 747 788 Mountain-300 Black, 48 BK-M47B-48 740 789 Road-250 Red, 44 BK-R89R-44 895 790 Road-250 Red, 48 BK-R89R-48 812 791 Road-250 Red, 52 BK-R89R-52 639 792 Road-250 Red, 58 BK-R89R-58 946 793 Road-250 Black, 44 BK-R89B-44 1642 794 Road-250 Black, 48 BK-R89B-48 1498 795 Road-250 Black, 52 BK-R89B-52 1245 796 Road-250 Black, 58 BK-R89B-58 910 797 Road-550-W Yellow, 38 BK-R64Y-38 1744 798 Road-550-W Yellow, 40 BK-R64Y-40 1500 799 Road-550-W Yellow, 42 BK-R64Y-42 1249 800 Road-550-W Yellow, 44 BK-R64Y-44 927 801 Road-550-W Yellow, 48 BK-R64Y-48 1763 802 LL Fork FK-1639 190 804 HL Fork FK-9939 444 805 LL Headset HS-0296 95 806 ML Headset HS-2451 659 807 HL Headset HS-3479 255 808 LL Mountain Handlebars HB-M243 865 809 ML Mountain Handlebars HB-M763 1465 810 HL Mountain Handlebars HB-M918 500 811 LL Road Handlebars HB-R504 213 813 HL Road Handlebars HB-R956 631 814 ML Mountain Frame - Black, 38 FR-M63B-38 514 815 LL Mountain Front Wheel FW-M423 617 816 ML Mountain Front Wheel FW-M762 


![sad panda](./DangerWillRobinson.gif "Figure 3 - Using a CROSS APPLY with the [fn_TotalSoldToDate] to return the total number sold for each productID")</br>


## There is a challenge with the query shown in Figure 3.  
## With the query in Figure 1 there was 295 rows returned.  
## The query in Figure 3 only returns 266 rows.  

## The reason for this is that the query is using **CROSS APPLY**, when the function is executed a record only be return if there match with the left table.  So if the function does not return a record for the ProductID ## no record will be returned in the recordset.

## Now let's use the **OUTER APPLY** keyword with the same query as used in Figure 3 and see what the differences are.


![sad panda](./Figure4.jpg "Using an OUTER APPLY with the [fn_TotalSoldToDate] to return the total number sold for each productID")</br>
Figure 4 - Using an OUTER APPLY with the [fn_TotalSoldToDate] to return the total number sold for each productID

In [6]:
SELECT [P].[ProductID],
       [P].[Name],
       [P].[ProductNumber],
	   [PS].[TotalOrdered]
FROM [Production].[Product] AS [P]
OUTER APPLY [dbo].[fn_TotalSoldToDate]([P].ProductID) AS [PS]
WHERE [P].[FinishedGoodsFlag] = 1;

(295 rows affected)

Total execution time: 00:00:00.061

ProductID Name ProductNumber TotalOrdered 680 HL Road Frame - Black, 58 FR-R92B-58 NULL 706 HL Road Frame - Red, 58 FR-R92R-58 NULL 707 Sport-100 Helmet, Red HL-U509-R 6266 708 Sport-100 Helmet, Black HL-U509 6532 709 Mountain Bike Socks, M SO-B909-M 1107 710 Mountain Bike Socks, L SO-B909-L 90 711 Sport-100 Helmet, Blue HL-U509-B 6743 712 AWC Logo Cap CA-1098 8311 713 Long-Sleeve Logo Jersey, S LJ-0192-S 429 714 Long-Sleeve Logo Jersey, M LJ-0192-M 3636 715 Long-Sleeve Logo Jersey, L LJ-0192-L 6592 716 Long-Sleeve Logo Jersey, XL LJ-0192-X 2980 717 HL Road Frame - Red, 62 FR-R92R-62 485 718 HL Road Frame - Red, 44 FR-R92R-44 486 719 HL Road Frame - Red, 48 FR-R92R-48 109 720 HL Road Frame - Red, 52 FR-R92R-52 NULL 721 HL Road Frame - Red, 56 FR-R92R-56 NULL 722 LL Road Frame - Black, 58 FR-R38B-58 940 723 LL Road Frame - Black, 60 FR-R38B-60 129 724 LL Road Frame - Black, 62 FR-R38B-62 NULL 725 LL Road Frame - Red, 44 FR-R38R-44 991 726 LL Road Frame - Red, 48 FR-R38R-48 671 727 LL Road Frame - Red, 52 FR-R38R-52 100 728 LL Road Frame - Red, 58 FR-R38R-58 NULL 729 LL Road Frame - Red, 60 FR-R38R-60 996 730 LL Road Frame - Red, 62 FR-R38R-62 697 731 ML Road Frame - Red, 44 FR-R72R-44 NULL 732 ML Road Frame - Red, 48 FR-R72R-48 250 733 ML Road Frame - Red, 52 FR-R72R-52 90 734 ML Road Frame - Red, 58 FR-R72R-58 NULL 735 ML Road Frame - Red, 60 FR-R72R-60 NULL 736 LL Road Frame - Black, 44 FR-R38B-44 238 737 LL Road Frame - Black, 48 FR-R38B-48 NULL 738 LL Road Frame - Black, 52 FR-R38B-52 1581 739 HL Mountain Frame - Silver, 42 FR-M94S-42 347 740 HL Mountain Frame - Silver, 44 FR-M94S-44 NULL 741 HL Mountain Frame - Silver, 48 FR-M94S-52 173 742 HL Mountain Frame - Silver, 46 FR-M94S-46 649 743 HL Mountain Frame - Black, 42 FR-M94B-42 1181 744 HL Mountain Frame - Black, 44 FR-M94B-44 17 745 HL Mountain Frame - Black, 48 FR-M94B-48 131 746 HL Mountain Frame - Black, 46 FR-M94B-46 NULL 747 HL Mountain Frame - Black, 38 FR-M94B-38 659 748 HL Mountain Frame - Silver, 38 FR-M94S-38 1206 749 Road-150 Red, 62 BK-R93R-62 600 750 Road-150 Red, 44 BK-R93R-44 437 751 Road-150 Red, 48 BK-R93R-48 493 752 Road-150 Red, 52 BK-R93R-52 458 753 Road-150 Red, 56 BK-R93R-56 664 754 Road-450 Red, 58 BK-R68R-58 582 755 Road-450 Red, 60 BK-R68R-60 350 756 Road-450 Red, 44 BK-R68R-44 346 757 Road-450 Red, 48 BK-R68R-48 156 758 Road-450 Red, 52 BK-R68R-52 710 759 Road-650 Red, 58 BK-R50R-58 618 760 Road-650 Red, 60 BK-R50R-60 2232 761 Road-650 Red, 62 BK-R50R-62 1896 762 Road-650 Red, 44 BK-R50R-44 2254 763 Road-650 Red, 48 BK-R50R-48 1888 764 Road-650 Red, 52 BK-R50R-52 1112 765 Road-650 Black, 58 BK-R50B-58 1871 766 Road-650 Black, 60 BK-R50B-60 1100 767 Road-650 Black, 62 BK-R50B-62 644 768 Road-650 Black, 44 BK-R50B-44 1080 769 Road-650 Black, 48 BK-R50B-48 644 770 Road-650 Black, 52 BK-R50B-52 2270 771 Mountain-100 Silver, 38 BK-M82S-38 642 772 Mountain-100 Silver, 42 BK-M82S-42 593 773 Mountain-100 Silver, 44 BK-M82S-44 601 774 Mountain-100 Silver, 48 BK-M82S-48 505 775 Mountain-100 Black, 38 BK-M82B-38 682 776 Mountain-100 Black, 42 BK-M82B-42 634 777 Mountain-100 Black, 44 BK-M82B-44 678 778 Mountain-100 Black, 48 BK-M82B-48 616 779 Mountain-200 Silver, 38 BK-M68S-38 2394 780 Mountain-200 Silver, 42 BK-M68S-42 2234 781 Mountain-200 Silver, 46 BK-M68S-46 2216 782 Mountain-200 Black, 38 BK-M68B-38 2977 783 Mountain-200 Black, 42 BK-M68B-42 2664 784 Mountain-200 Black, 46 BK-M68B-46 2111 785 Mountain-300 Black, 38 BK-M47B-38 684 786 Mountain-300 Black, 40 BK-M47B-40 776 787 Mountain-300 Black, 44 BK-M47B-44 747 788 Mountain-300 Black, 48 BK-M47B-48 740 789 Road-250 Red, 44 BK-R89R-44 895 790 Road-250 Red, 48 BK-R89R-48 812 791 Road-250 Red, 52 BK-R89R-52 639 792 Road-250 Red, 58 BK-R89R-58 946 793 Road-250 Black, 44 BK-R89B-44 1642 794 Road-250 Black, 48 BK-R89B-48 1498 795 Road-250 Black, 52 BK-R89B-52 1245 796 Road-250 Black, 58 BK-R89B-58 910 797 Road-550-W Yellow, 38 BK-R64Y-38 1744 798 Road-550-W Yellow, 40 BK-R64Y-40 1500 799 Road-550-W Yell

## Using OUTER APPLY the results set returns all the records from the left hand table.  That can be seen from the fact we have 295 records returned by the query in Figure 4.  Also it's worthwhile noting that for ProductID’s 680 and 706, the field [TotalOrdered] returns NULL, since these products have never been sold.

## **Now we have finished introducing the APPLY keyword let's look at some other uses.  Whilst the APPLY keyword was intended to be used with Table Valued Functions.  There are a number of other ways to use the APPLY word.**
***


# Calling scalar functions

![sad panda](./Figure5.jpg "Figure 5 - Using a CROSS APPLY to call the function [ufnGetStock]")</br>
Figure 5 - Using a CROSS APPLY to call the function [ufnGetStock]

In [7]:
/*Method 1 - Inline function call*/

SELECT [P].[ProductID],
       [P].[Name],
       [P].[ProductNumber],
	   [dbo].[ufnGetStock]([P].[ProductID]) AS [StockLevel]
FROM [Production].[Product] AS [P]
WHERE [P].[FinishedGoodsFlag] = 1;

/*Method 2 - Call function using cross apply*/

SELECT [P].[ProductID],
       [P].[Name],
       [P].[ProductNumber],
	   [Stock].[Level]
FROM [Production].[Product] AS [P]
CROSS APPLY 
(SELECT [dbo].[ufnGetStock]([P].[ProductID]) AS [Level]) AS [Stock]
WHERE [P].[FinishedGoodsFlag] = 1;


(295 rows affected)

(295 rows affected)

Total execution time: 00:00:00.079

ProductID Name ProductNumber StockLevel 680 HL Road Frame - Black, 58 FR-R92B-58 0 706 HL Road Frame - Red, 58 FR-R92R-58 0 707 Sport-100 Helmet, Red HL-U509-R 0 708 Sport-100 Helmet, Black HL-U509 0 709 Mountain Bike Socks, M SO-B909-M 0 710 Mountain Bike Socks, L SO-B909-L 0 711 Sport-100 Helmet, Blue HL-U509-B 0 712 AWC Logo Cap CA-1098 0 713 Long-Sleeve Logo Jersey, S LJ-0192-S 0 714 Long-Sleeve Logo Jersey, M LJ-0192-M 0 715 Long-Sleeve Logo Jersey, L LJ-0192-L 0 716 Long-Sleeve Logo Jersey, XL LJ-0192-X 0 717 HL Road Frame - Red, 62 FR-R92R-62 0 718 HL Road Frame - Red, 44 FR-R92R-44 0 719 HL Road Frame - Red, 48 FR-R92R-48 0 720 HL Road Frame - Red, 52 FR-R92R-52 0 721 HL Road Frame - Red, 56 FR-R92R-56 0 722 LL Road Frame - Black, 58 FR-R38B-58 0 723 LL Road Frame - Black, 60 FR-R38B-60 0 724 LL Road Frame - Black, 62 FR-R38B-62 0 725 LL Road Frame - Red, 44 FR-R38R-44 0 726 LL Road Frame - Red, 48 FR-R38R-48 0 727 LL Road Frame - Red, 52 FR-R38R-52 0 728 LL Road Frame - Red, 58 FR-R38R-58 0 729 LL Road Frame - Red, 60 FR-R38R-60 0 730 LL Road Frame - Red, 62 FR-R38R-62 0 731 ML Road Frame - Red, 44 FR-R72R-44 0 732 ML Road Frame - Red, 48 FR-R72R-48 0 733 ML Road Frame - Red, 52 FR-R72R-52 0 734 ML Road Frame - Red, 58 FR-R72R-58 0 735 ML Road Frame - Red, 60 FR-R72R-60 0 736 LL Road Frame - Black, 44 FR-R38B-44 0 737 LL Road Frame - Black, 48 FR-R38B-48 0 738 LL Road Frame - Black, 52 FR-R38B-52 0 739 HL Mountain Frame - Silver, 42 FR-M94S-42 0 740 HL Mountain Frame - Silver, 44 FR-M94S-44 0 741 HL Mountain Frame - Silver, 48 FR-M94S-52 0 742 HL Mountain Frame - Silver, 46 FR-M94S-46 0 743 HL Mountain Frame - Black, 42 FR-M94B-42 0 744 HL Mountain Frame - Black, 44 FR-M94B-44 0 745 HL Mountain Frame - Black, 48 FR-M94B-48 0 746 HL Mountain Frame - Black, 46 FR-M94B-46 0 747 HL Mountain Frame - Black, 38 FR-M94B-38 0 748 HL Mountain Frame - Silver, 38 FR-M94S-38 0 749 Road-150 Red, 62 BK-R93R-62 0 750 Road-150 Red, 44 BK-R93R-44 0 751 Road-150 Red, 48 BK-R93R-48 0 752 Road-150 Red, 52 BK-R93R-52 0 753 Road-150 Red, 56 BK-R93R-56 0 754 Road-450 Red, 58 BK-R68R-58 0 755 Road-450 Red, 60 BK-R68R-60 0 756 Road-450 Red, 44 BK-R68R-44 0 757 Road-450 Red, 48 BK-R68R-48 0 758 Road-450 Red, 52 BK-R68R-52 0 759 Road-650 Red, 58 BK-R50R-58 0 760 Road-650 Red, 60 BK-R50R-60 0 761 Road-650 Red, 62 BK-R50R-62 0 762 Road-650 Red, 44 BK-R50R-44 0 763 Road-650 Red, 48 BK-R50R-48 0 764 Road-650 Red, 52 BK-R50R-52 0 765 Road-650 Black, 58 BK-R50B-58 0 766 Road-650 Black, 60 BK-R50B-60 0 767 Road-650 Black, 62 BK-R50B-62 0 768 Road-650 Black, 44 BK-R50B-44 0 769 Road-650 Black, 48 BK-R50B-48 0 770 Road-650 Black, 52 BK-R50B-52 0 771 Mountain-100 Silver, 38 BK-M82S-38 0 772 Mountain-100 Silver, 42 BK-M82S-42 0 773 Mountain-100 Silver, 44 BK-M82S-44 0 774 Mountain-100 Silver, 48 BK-M82S-48 0 775 Mountain-100 Black, 38 BK-M82B-38 0 776 Mountain-100 Black, 42 BK-M82B-42 0 777 Mountain-100 Black, 44 BK-M82B-44 0 778 Mountain-100 Black, 48 BK-M82B-48 0 779 Mountain-200 Silver, 38 BK-M68S-38 0 780 Mountain-200 Silver, 42 BK-M68S-42 0 781 Mountain-200 Silver, 46 BK-M68S-46 0 782 Mountain-200 Black, 38 BK-M68B-38 0 783 Mountain-200 Black, 42 BK-M68B-42 0 784 Mountain-200 Black, 46 BK-M68B-46 0 785 Mountain-300 Black, 38 BK-M47B-38 0 786 Mountain-300 Black, 40 BK-M47B-40 0 787 Mountain-300 Black, 44 BK-M47B-44 0 788 Mountain-300 Black, 48 BK-M47B-48 0 789 Road-250 Red, 44 BK-R89R-44 0 790 Road-250 Red, 48 BK-R89R-48 0 791 Road-250 Red, 52 BK-R89R-52 0 792 Road-250 Red, 58 BK-R89R-58 0 793 Road-250 Black, 44 BK-R89B-44 0 794 Road-250 Black, 48 BK-R89B-48 0 795 Road-250 Black, 52 BK-R89B-52 0 796 Road-250 Black, 58 BK-R89B-58 0 797 Road-550-W Yellow, 38 BK-R64Y-38 0 798 Road-550-W Yellow, 40 BK-R64Y-40 0 799 Road-550-W Yellow, 42 BK-R64Y-42 0 800 Road-550-W Yellow, 44 BK-R64Y-44 0 801 Road-550-W Yellow, 48 BK-R64Y-48 0 802 LL Fork FK-1639 0 803 ML Fork FK-5136 0 804 HL Fork FK-9939 0 805 LL Headset HS-0296 265 806 ML Headset HS-2451 268 807 HL Heads

ProductID Name ProductNumber Level 680 HL Road Frame - Black, 58 FR-R92B-58 0 706 HL Road Frame - Red, 58 FR-R92R-58 0 707 Sport-100 Helmet, Red HL-U509-R 0 708 Sport-100 Helmet, Black HL-U509 0 709 Mountain Bike Socks, M SO-B909-M 0 710 Mountain Bike Socks, L SO-B909-L 0 711 Sport-100 Helmet, Blue HL-U509-B 0 712 AWC Logo Cap CA-1098 0 713 Long-Sleeve Logo Jersey, S LJ-0192-S 0 714 Long-Sleeve Logo Jersey, M LJ-0192-M 0 715 Long-Sleeve Logo Jersey, L LJ-0192-L 0 716 Long-Sleeve Logo Jersey, XL LJ-0192-X 0 717 HL Road Frame - Red, 62 FR-R92R-62 0 718 HL Road Frame - Red, 44 FR-R92R-44 0 719 HL Road Frame - Red, 48 FR-R92R-48 0 720 HL Road Frame - Red, 52 FR-R92R-52 0 721 HL Road Frame - Red, 56 FR-R92R-56 0 722 LL Road Frame - Black, 58 FR-R38B-58 0 723 LL Road Frame - Black, 60 FR-R38B-60 0 724 LL Road Frame - Black, 62 FR-R38B-62 0 725 LL Road Frame - Red, 44 FR-R38R-44 0 726 LL Road Frame - Red, 48 FR-R38R-48 0 727 LL Road Frame - Red, 52 FR-R38R-52 0 728 LL Road Frame - Red, 58 FR-R38R-58 0 729 LL Road Frame - Red, 60 FR-R38R-60 0 730 LL Road Frame - Red, 62 FR-R38R-62 0 731 ML Road Frame - Red, 44 FR-R72R-44 0 732 ML Road Frame - Red, 48 FR-R72R-48 0 733 ML Road Frame - Red, 52 FR-R72R-52 0 734 ML Road Frame - Red, 58 FR-R72R-58 0 735 ML Road Frame - Red, 60 FR-R72R-60 0 736 LL Road Frame - Black, 44 FR-R38B-44 0 737 LL Road Frame - Black, 48 FR-R38B-48 0 738 LL Road Frame - Black, 52 FR-R38B-52 0 739 HL Mountain Frame - Silver, 42 FR-M94S-42 0 740 HL Mountain Frame - Silver, 44 FR-M94S-44 0 741 HL Mountain Frame - Silver, 48 FR-M94S-52 0 742 HL Mountain Frame - Silver, 46 FR-M94S-46 0 743 HL Mountain Frame - Black, 42 FR-M94B-42 0 744 HL Mountain Frame - Black, 44 FR-M94B-44 0 745 HL Mountain Frame - Black, 48 FR-M94B-48 0 746 HL Mountain Frame - Black, 46 FR-M94B-46 0 747 HL Mountain Frame - Black, 38 FR-M94B-38 0 748 HL Mountain Frame - Silver, 38 FR-M94S-38 0 749 Road-150 Red, 62 BK-R93R-62 0 750 Road-150 Red, 44 BK-R93R-44 0 751 Road-150 Red, 48 BK-R93R-48 0 752 Road-150 Red, 52 BK-R93R-52 0 753 Road-150 Red, 56 BK-R93R-56 0 754 Road-450 Red, 58 BK-R68R-58 0 755 Road-450 Red, 60 BK-R68R-60 0 756 Road-450 Red, 44 BK-R68R-44 0 757 Road-450 Red, 48 BK-R68R-48 0 758 Road-450 Red, 52 BK-R68R-52 0 759 Road-650 Red, 58 BK-R50R-58 0 760 Road-650 Red, 60 BK-R50R-60 0 761 Road-650 Red, 62 BK-R50R-62 0 762 Road-650 Red, 44 BK-R50R-44 0 763 Road-650 Red, 48 BK-R50R-48 0 764 Road-650 Red, 52 BK-R50R-52 0 765 Road-650 Black, 58 BK-R50B-58 0 766 Road-650 Black, 60 BK-R50B-60 0 767 Road-650 Black, 62 BK-R50B-62 0 768 Road-650 Black, 44 BK-R50B-44 0 769 Road-650 Black, 48 BK-R50B-48 0 770 Road-650 Black, 52 BK-R50B-52 0 771 Mountain-100 Silver, 38 BK-M82S-38 0 772 Mountain-100 Silver, 42 BK-M82S-42 0 773 Mountain-100 Silver, 44 BK-M82S-44 0 774 Mountain-100 Silver, 48 BK-M82S-48 0 775 Mountain-100 Black, 38 BK-M82B-38 0 776 Mountain-100 Black, 42 BK-M82B-42 0 777 Mountain-100 Black, 44 BK-M82B-44 0 778 Mountain-100 Black, 48 BK-M82B-48 0 779 Mountain-200 Silver, 38 BK-M68S-38 0 780 Mountain-200 Silver, 42 BK-M68S-42 0 781 Mountain-200 Silver, 46 BK-M68S-46 0 782 Mountain-200 Black, 38 BK-M68B-38 0 783 Mountain-200 Black, 42 BK-M68B-42 0 784 Mountain-200 Black, 46 BK-M68B-46 0 785 Mountain-300 Black, 38 BK-M47B-38 0 786 Mountain-300 Black, 40 BK-M47B-40 0 787 Mountain-300 Black, 44 BK-M47B-44 0 788 Mountain-300 Black, 48 BK-M47B-48 0 789 Road-250 Red, 44 BK-R89R-44 0 790 Road-250 Red, 48 BK-R89R-48 0 791 Road-250 Red, 52 BK-R89R-52 0 792 Road-250 Red, 58 BK-R89R-58 0 793 Road-250 Black, 44 BK-R89B-44 0 794 Road-250 Black, 48 BK-R89B-48 0 795 Road-250 Black, 52 BK-R89B-52 0 796 Road-250 Black, 58 BK-R89B-58 0 797 Road-550-W Yellow, 38 BK-R64Y-38 0 798 Road-550-W Yellow, 40 BK-R64Y-40 0 799 Road-550-W Yellow, 42 BK-R64Y-42 0 800 Road-550-W Yellow, 44 BK-R64Y-44 0 801 Road-550-W Yellow, 48 BK-R64Y-48 0 802 LL Fork FK-1639 0 803 ML Fork FK-5136 0 804 HL Fork FK-9939 0 805 LL Headset HS-0296 265 806 ML Headset HS-2451 268 807 HL Headset HS

## As you can see there is just two different ways of calling the same scalar function.  

## With this particular example there is not different in the performance, the query optimiser is far too clever for us in this case.  

## Can you do other things with the APPLY keyword, well yes I am so glad you asked

## Let's take an example we want to return all the last three orders from each sales person in table [Sales].[SalesOrderHeader].  Ordered by [OrderDate] field</br>

![sad panda](./Figure6.jpg "Figure 6 - Showing one method for solving the question posed above")</br>
Figure 6 - Using a CROSS APPLY to call the function [ufnGetStock]

## As shown in Figure 6 the inner query adds a row number of the query, and excludes any records which do not have a SalesPersonID.  The outer query then has to filter the resulting dataset to only return the first three records for each SalesPersonID returned by the inner query.  This pattern could easily get more complicated and harder to read.  Next let's see if can write the same query using CROSS APPLY.

![sad panda](./Figure7.jpg "Figure 7 - Rewritten query using CROSS APPLY")</br>
Figure 7 - Rewritten query using CROSS APPLY

## With the example shown in Figure 7, the entire sql query is contained within the CROSS APPLY.  One difference between the query in Figure 6 and Figure 7 is readability.

# Why use CROSS APPLY or OUTER APPLY

## For me its about whats the easiest way to get answers.  Let's take an example of a query for CCF which wanted to return same values from Classifications and Grant Contact details

## For each Grant Record 
## Grant Classifications - return 1st speciality in a column, 2nd specialty in a different column
## For the Trainee contact type, return work Email address and work phone number.


![sad panda](./Figure8.jpg "Figure 8 -  Query for TCC showing the entire sample query")</br>
Figure 8 -  Query for TCC showing the entire sample query

# Let's look section 1 first

![sad panda](./Figure8-section1.jpg "Figure 8.1 - the section of the query returning specific Grant Contact Details")</br>
Figure 8.1 - the section of the query returning specific Grant Contact Details

## First thing to note is an **OUTER APPLY** is being used, the reason for this is that even if the query within the **OUTER APPLY** returns zero records.  The other side of the join will always return records.  This could have been done using a **LEFT JOIN**, with the predicates (the where clause) as part of the join.  Hence this is same more about coding style.

![sad panda](./Figure8-section2.jpg "Figure 8.2 - the section of the query returning specific Grant Classifications")</br>
Figure 8.2 - the section of the query returning specific Grant Classifications

## The same principle as before this could have been done using a correlated subquery with a **LEFT JOIN**. The use if the **OUTER APPLY** is more about readability and maintainability.  If require to add more classifications, then copy the query for one update the WHERE clause and name and the work done.

# Sometimes you cannot use a JOIN

![sad panda](./frightened.gif "Oh no cannot use a join")</br>

## Let try and run this code to get some information from the system DMV's


In [8]:
SELECT * FROM SYS.dm_exec_cached_plans CP
INNER JOIN SYS.dm_exec_sql_text(CP.plan_handle) CA

: Msg 102, Level 15, State 1, Line 2
Incorrect syntax near 'CA'.

Total execution time: 00:00:00

# CROSS APPLY 
![sad panda](./rescue.jpg "Oh no cannot use a join")</br>

In [9]:
SELECT * FROM SYS.dm_exec_cached_plans CP
CROSS APPLY SYS.dm_exec_sql_text(CP.plan_handle) CA

(53 rows affected)

Total execution time: 00:00:00.038

bucketid refcounts usecounts size_in_bytes memory_object_address cacheobjtype objtype plan_handle pool_id parent_plan_handle dbid objectid number encrypted text 10245 2 1 57344 0x000001BB9DC76060 Compiled Plan Adhoc 0x06000B0025F7DF328002D2ACBB01000001000000000000000000000000000000000000000000000000000000 2 NULL 11 NULL NULL 0 SELECT * FROM SYS.dm_exec_cached_plans CP
CROSS APPLY SYS.dm_exec_sql_text(CP.plan_handle) CA 31491 2 1 368640 0x000001BB52C92060 Compiled Plan Prepared 0x06000B0056CDA837A0D5659BBB01000001000000000000000000000000000000000000000000000000000000 2 NULL 11 NULL NULL 0 (@_msparam_0 nvarchar(4000),@_msparam_1 nvarchar(4000),@_msparam_2 nvarchar(4000))SELECT
clmns.column_id AS [ID],
clmns.name AS [Name],
ISNULL(dc.Name, N'') AS [DefaultConstraintName],
clmns.is_nullable AS [Nullable],
CAST(ISNULL(cik.index_column_id, 0) AS bit) AS [InPrimaryKey],
clmns.is_identity AS [Identity],
usrt.name AS [DataType],
ISNULL(baset.name, N'') AS [SystemType],
CAST(CASE WHEN baset.name IN (N'nchar', N'nvarchar') AND clmns.max_length <> -1 THEN clmns.max_length/2 ELSE clmns.max_length END AS int) AS [Length],
CAST(clmns.precision AS int) AS [NumericPrecision],
CAST(clmns.scale AS int) AS [NumericScale],
ISNULL(xscclmns.name, N'') AS [XmlSchemaNamespace],
ISNULL(s2clmns.name, N'') AS [XmlSchemaNamespaceSchema],
ISNULL( (case clmns.is_xml_document when 1 then 2 else 1 end), 0) AS [XmlDocumentConstraint],
s1clmns.name AS [DataTypeSchema],
clmns.is_computed AS [Computed]
FROM
sys.all_views AS v
INNER JOIN sys.all_columns AS clmns ON clmns.object_id=v.object_id
LEFT OUTER JOIN sys.default_constraints as dc ON clmns.default_object_id = dc.object_id
LEFT OUTER JOIN sys.indexes AS ik ON ik.object_id = clmns.object_id and 1=ik.is_primary_key
LEFT OUTER JOIN sys.index_columns AS cik ON cik.index_id = ik.index_id and cik.column_id = clmns.column_id and cik.object_id = clmns.object_id and 0 = cik.is_included_column
LEFT OUTER JOIN sys.types AS usrt ON usrt.user_type_id = clmns.user_type_id
LEFT OUTER JOIN sys.types AS baset ON (baset.user_type_id = clmns.system_type_id and baset.user_type_id = baset.system_type_id) or ((baset.system_type_id = clmns.system_type_id) and (baset.user_type_id = clmns.user_type_id) and (baset.is_user_defined = 0) and (baset.is_assembly_type = 1)) 
LEFT OUTER JOIN sys.xml_schema_collections AS xscclmns ON xscclmns.xml_collection_id = clmns.xml_collection_id
LEFT OUTER JOIN sys.schemas AS s2clmns ON s2clmns.schema_id = xscclmns.schema_id
LEFT OUTER JOIN sys.schemas AS s1clmns ON s1clmns.schema_id = usrt.schema_id
WHERE
(v.type = @_msparam_0)and(v.name=@_msparam_1 and SCHEMA_NAME(v.schema_id)=@_msparam_2)
ORDER BY
[ID] ASC 36249 2 1 376832 0x000001BB52C96060 Compiled Plan Prepared 0x06000B00D33AFA0620CE659BBB01000001000000000000000000000000000000000000000000000000000000 2 NULL 11 NULL NULL 0 (@_msparam_0 nvarchar(4000),@_msparam_1 nvarchar(4000))SELECT
clmns.column_id AS [ID],
clmns.name AS [Name],
ISNULL(dc.Name, N'') AS [DefaultConstraintName],
clmns.is_nullable AS [Nullable],
CAST(ISNULL(cik.index_column_id, 0) AS bit) AS [InPrimaryKey],
clmns.is_identity AS [Identity],
usrt.name AS [DataType],
ISNULL(baset.name, N'') AS [SystemType],
CAST(CASE WHEN baset.name IN (N'nchar', N'nvarchar') AND clmns.max_length <> -1 THEN clmns.max_length/2 ELSE clmns.max_length END AS int) AS [Length],
CAST(clmns.precision AS int) AS [NumericPrecision],
CAST(clmns.scale AS int) AS [NumericScale],
ISNULL(xscclmns.name, N'') AS [XmlSchemaNamespace],
ISNULL(s2clmns.name, N'') AS [XmlSchemaNamespaceSchema],
ISNULL( (case clmns.is_xml_document when 1 then 2 else 1 end), 0) AS [XmlDocumentConstraint],
s1clmns.name AS [DataTypeSchema],
clmns.is_computed AS [Computed]
FROM
sys.all_objects AS udf
INNER JOIN sys.all_columns AS clmns ON clmns.object_id=udf.object_id
LEFT OUTER JOIN sys.default_constraints as dc ON clmns.default_object_id = dc.object_id
LEFT OUTER JOIN sys.indexes AS ik ON ik.object_id = clmns.object_id and 1=ik.is_primary_ke

https://explainextended.com/2009/07/16/inner-join-vs-cross-apply/

https://www.mssqltips.com/sqlservertip/1958/sql-server-cross-apply-and-outer-apply/

https://www.codeproject.com/Articles/607246/Making-OUTER-and-CROSS-APPLY-work-for-you
https://www.red-gate.com/simple-talk/sql/t-sql-programming/sql-server-apply-basics/


![sad panda](./GrumpyCat_Not_Impressed.jpg "Figure 8.2 - the section of the query returning specific Grant Classifications")</br>
# **Grumpy Cat is not impressed**
# **Are you impressed??????**